torchtitan/experiments/qwen3/model/model.py

```diff
@@ -0,0 +1,454 @@
+        xq.view(*xq.shape[:-1], 2, xq.shape[-1] // 2)
+        .transpose(-2, -1)
+        .contiguous()
+        .float()
+    )
+    freqs_cis = reshape_for_broadcast(freqs_cis, xq_complex)
+
+    xq_out = torch.view_as_real(xq_complex * freqs_cis).flatten(3)
+    xk_out = torch.view_as_real(xk_complex * freqs_cis).flatten(3)
+
+    return xq_out.type_as(xq), xk_out.type_as(xk)
+
+
+def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
+    """torch.repeat_interleave(x, dim=2, repeats=n_rep)"""
+    bs, slen, n_kv_heads, head_dim = x.shape
+    if n_rep == 1:
+        return x
+    return (
+        torch.unsqueeze(x, dim=3)
+        .expand(bs, slen, n_kv_heads, n_rep, head_dim)
+        .reshape(bs, slen, n_kv_heads * n_rep, head_dim)
+    )
+
+
+class Attention(nn.Module):
+    """
+    Multi-head attention module.
+
+    Args:
+        model_args (TransformerModelArgs): Model configuration arguments.
+
+    Attributes:
+        n_kv_heads (int): Number of key and value heads.
+        n_heads (int): Number of query heads.
+        n_rep (int): Number of repetitions for local heads.
+        head_dim (int): Dimension size of each attention head.
+        wq (Linear): Linear transformation for queries.
+        wk (Linear): Linear transformation for keys.
+        wv (Linear): Linear transformation for values.
+        wo (Linear): Linear transformation for output.
+
+    """
+
+    def __init__(self, model_args: Qwen3ModelArgs):
+        super().__init__()
+        self.n_heads = model_args.n_heads
+        self.n_kv_heads = (
+            model_args.n_heads
+            if model_args.n_kv_heads is None
+            else model_args.n_kv_heads
+        )
+        self.n_rep = self.n_heads // self.n_kv_heads
+        self.head_dim = model_args.head_dim
+
+        # RMSNorm added here to the here to include the q-k norm
+        # This is one of the main differences between Llama3 and Qwen3
+        if model_args.qk_norm:
+            self.q_norm = nn.RMSNorm(
+                self.head_dim, eps=model_args.norm_eps, elementwise_affine=True
+            )
+            self.k_norm = nn.RMSNorm(
+                self.head_dim, eps=model_args.norm_eps, elementwise_affine=True
+            )
+
+        self.wq = nn.Linear(
+            model_args.dim, model_args.n_heads * self.head_dim, bias=False
+        )
+        self.wk = nn.Linear(model_args.dim, self.n_kv_heads * self.head_dim, bias=False)
+        self.wv = nn.Linear(model_args.dim, self.n_kv_heads * self.head_dim, bias=False)
+        self.wo = nn.Linear(
+            model_args.n_heads * self.head_dim, model_args.dim, bias=False
+        )
+        self.sdpa = build_attention(model_args.use_flex_attn, model_args.attn_mask_type)
+
+    def init_weights(self, init_std: float):
+        for linear in (self.wq, self.wk, self.wv):
+            nn.init.trunc_normal_(linear.weight, mean=0.0, std=0.02)
+        nn.init.trunc_normal_(self.wo.weight, mean=0.0, std=init_std)
```

Comment:
We need to init q_norm , k_norm as well

Reply:
eg, other norm layers are initialized as:
if self.norm is not None:
            self.norm.reset_parameters()